crates/ruma-api/src/lib.rs

```diff
@@ -591,3 +595,82 @@ impl MatrixVersion {
         }
     }
 }
+
+/// A specifier for what path variant of an endpoint to request.
+#[derive(Clone, PartialEq, Eq, Debug)]
+#[allow(clippy::exhaustive_enums)]
+pub enum EndpointPath {
+    /// Select an appropriate path version for a particular matrix version.
+    ///
+    /// Note: When using
+    /// [`try_into_http_request`](OutgoingRequest::try_into_http_request) with this variant, it can
+    /// emit additional errors;
+    /// - [`NoUnstablePath`](IntoHttpError::NoUnstablePath), if the matrix version is too new, and
+    ///   the endpoint has no defined `unstable` path (for maintenance or legacy reasons.)
+    /// - [`EndpointRemoved`](IntoHttpError::EndpointRemoved), if the matrix version is too old,
+    ///   and the endpoint is known to be removed in the same or a newer version.
+    ForMatrixVersion(MatrixVersion),
+
+    /// Select any path, preferring stable.
+    ///
+    /// Note: This does not emit any errors like
+    /// [`ForMatrixVersion`](EndpointPath::ForMatrixVersion) does, if the server's lowest matrix
+    /// version has removed the endpoint, or it has not yet stabilized the endpoint, this may
+    /// result in runtime HTTP errors, it is highly recommended to fetch `/versions`, extract
+    /// Ruma-known matrix versions, and pass the preferred version through `ForMatrixVersion`.
+    PreferStable,
+}
+
+impl From<MatrixVersion> for EndpointPath {
+    fn from(ver: MatrixVersion) -> Self {
+        Self::ForMatrixVersion(ver)
+    }
+}
+
+// This function needs to be public, yet hidden, as all `try_into_http_request`s would be using it.
+// We're also using MatrixVersion::repr() here, which is crate-private.
+#[doc(hidden)]
+pub fn select_path<'a>(
+    path: EndpointPath,
+    metadata: &'_ Metadata,
+    unstable: Option<fmt::Arguments<'a>>,
+    r0: Option<fmt::Arguments<'a>>,
+    stable: Option<fmt::Arguments<'a>>,
+) -> Result<fmt::Arguments<'a>, IntoHttpError> {
+    let version = match path {
+        EndpointPath::ForMatrixVersion(v) => v,
+        EndpointPath::PreferStable => {
+            return Ok(stable.or(r0).or(unstable).expect("one of three paths to be defined"))
+        }
+    };
+
+    if let Some(removed_ver) = metadata.removed {
+        if version.repr() >= removed_ver.repr() {
+            return Err(IntoHttpError::EndpointRemoved(version, removed_ver));
```

Comment:
I'd like this to simply log a warning, not error out. Think about the scenario of the SDK selecting the latest matrix version known to Ruma and calling a removed API endpoint – because we only know of one version that the server supports here, we don't know whether the endpoint is actually available or not.

Reply:
We do, because we're explicitly given a pointer to a version where this endpoint does not exist.

I think the opposite is a bigger footgun, where we log a warning but allow an application to not be able to deal with an unexpected 404 (as applications built from a certain time suddenly then aren't able to work with an endpoint).

We can check for this with this metadata, and thus we do so, and we also force the caller to deal with the error that could come from it (in one way or another).

Reply:
How would this new error be more useful than a 404?

> We do, because we're explicitly given a pointer to a version where this endpoint does not exist.

Well, I'm working off the assumption that you would generally pass "latest 1.0 compatible version that came from `get_supported_versions`" here, which to me seems like what we want to ensure unstable paths aren't relied upon for too long. The current logic would then error for removed (in latest spec) endpoints though even if they're still supported.

Reply:
Define "still supported", I do think that "picking the latest version unconditionally" is not the right way an application should do.

Perhaps we can add another variant, `ForMatrixVersions(&[MatrixVersion])`, from which this code will only error if *none* of the versions in that slice would support that endpoint.

I do want a check to make this invariant sure, 404 can mean both "this endpoint does not exist" and "this resource does not exist", which can have application logic consequences. At the very least with this approach, application logic has a branch for "this endpoint does not exist (for this server)" handled separately. (For comparison, this is equivalent to `Result<Option<T>, E>` instead of `Option<T>` the latter not telling if something failed, or if it doesn't exist.)

I do want to add a `ForMatrixVersions` variant, but I'm very much not in favour of removing this error branch.